Add resource aws_cloudwatch_event_bus

alexanderkalach opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#pull-requests --->

<!--- Please keep this note for the community --->

### Community Note
* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Relates #9330

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
* **New Resource:** `aws_cloudwatch_event_bus`
```

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSCloudWatchEventBus'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCloudWatchEventBus -timeout 120m
=== RUN   TestAccAWSCloudWatchEventBus_basic
=== PAUSE TestAccAWSCloudWatchEventBus_basic
=== RUN   TestAccAWSCloudWatchEventBus_disappears
=== PAUSE TestAccAWSCloudWatchEventBus_disappears
=== CONT  TestAccAWSCloudWatchEventBus_basic
=== CONT  TestAccAWSCloudWatchEventBus_disappears
--- PASS: TestAccAWSCloudWatchEventBus_disappears (33.66s)
--- PASS: TestAccAWSCloudWatchEventBus_basic (72.56s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	72.599s

$ make testacc TEST=./aws TESTARGS='-run=TestCloudWatchEventBusName'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestCloudWatchEventBusName -timeout 120m
=== RUN   TestCloudWatchEventBusName
--- PASS: TestCloudWatchEventBusName (0.00s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	0.027s
```

kurtmaile commented:
Hi,

What is the current state of this?  Very keen to try it out - thanks for the feature!

Cheers

alexanderkalach commented:
@kurtmaile conflicts are resolved. It's ready for review.

kurtmaile commented:
Great thanks sorry for delay getting back - seems to have passed and is merged.  Assume then its available in the next release?  Cheers!

Vlaaaaaaad commented:
Anything blocking this besides the merge conflicts? Can I help push this forward somehow?

fmasuhr commented on aws/validators.go:
```diff
@@ -2407,3 +2407,56 @@ func validateRoute53ResolverName(v interface{}, k string) (ws []string, errors [
+		errors = append(errors, fmt.Errorf("%q cannot be less than 1 character: %q", k, value))
+	} else if len(value) > 256 {
+		errors = append(errors, fmt.Errorf("%q cannot be longer than 256 characters: %q", k, value))
+	} else if value == "default" {
```
I tried this PR and used it to specify the `event_bus_name` attribute on `aws_cloudwatch_event_rule` and `aws_cloudwatch_event_target` but this check for `default` leads to an issue. 
The `event_bus_name` `default` does not need to be specified but when you omit this it will be added to the statefile because the AWS API contains it in the response. This will then lead to a change on the next apply as it would be changed from `default` to `null`.
As this validation forbids to specify `default` it currently is not working.

I suggest to have different validations for the `event_bus_name` attribute on `aws_cloudwatch_event_rule` and `aws_cloudwatch_event_target`  and for the `name` attribute on the `aws_cloudwatch_event_bus`

alexanderkalach replied:
Good catch! I would leave the current validation though. The issue was that `event_bus_name` in the statefile was updated with a value from AWS API. I think `event_bus_name` is kinda part of the identification for `aws_cloudwatch_event_rule` and `aws_cloudwatch_event_target`, so it shouldn't be updated. At least that's how I fixed now. Otherwise I think that `event_bus_name` should be a required attribute. But I prefer as an optional attribute to not break current configurations. The idea is that we should always omit `event_bus_name` if we want to use the `default` event bus.
Also I changed the import for `aws_cloudwatch_event_rule` and `aws_cloudwatch_event_target`. I use `#` instead of `/`, eg `event-bus-name#rule-name#target-id` or `rule-name#target-id` for the default event bus, because an event bus name can contain `/`, it's possible to detect such names, since all of them start with `aws.partner/`. Not sure, if it's better to use `/` or `#` for imports here 🤔

alexanderkalach replied:
I reverted the some changes, so `#` is not used to import `aws_cloudwatch_event_rule` and `aws_cloudwatch_event_target`

barneyparker commented:
Hi,  Any update in this one?  EventBridge support would be awesome, and it looks like the PR is in a pretty good state - any chance of a (rough) ETA on this getting merged?

ewbankkit commented:
@alexanderkalach As it will not be possible to create partner event buses (you have to be a SaaS provider, not just a regular AWS customer) with Terraform (currently) I suggest removing the `event_source_name` (and `policy`) attributes as these apply only to partner event buses.
This should allow the code to be simplified somewhat.

alexanderkalach commented:
@ewbankkit good point, thanks! I updated the PR.

ewbankkit commented on aws/resource_aws_cloudwatch_event_bus.go:
```diff
@@ -0,0 +1,91 @@
+	}
+
+	output, err := conn.DescribeEventBus(input)
+	if err != nil {
```
We need to handle the case where the resource has been deleted outside Terraform by checking for a _NotFound_ error, maybe:

```go
if isAWSErr(err, events.ErrCodeResourceNotFoundException, "") {
	log.Printf("[WARN] CloudWatch Event Bus (%s) not found, removing from state", d.Id())
	d.SetId("")
	return nil
}
```

other errors can be wrapped:

```go
if err != nil {
	return fmt.Errorf("error reading CloudWatch Event Bus: %s", err)
}
```

This can be tested by adding a `_disappears` test case, for example:
https://github.com/terraform-providers/terraform-provider-aws/blob/4875ae5fef0990f4a7e3768c81fcc732b90e0f13/aws/resource_aws_cloudwatch_log_stream_test.go#L33-L52

ewbankkit commented on aws/resource_aws_cloudwatch_event_bus_test.go:
```diff
@@ -0,0 +1,184 @@
+	return nil
+}
+
+func testAccCheckCloudWatchEventBusExists(n string) resource.TestCheckFunc {
```
May want to add a 2nd parameter `v *events.DescribeEventBusOutput` and set it (after the `resp == nil` check):

```go
*v = *resp
```

ewbankkit commented on aws/resource_aws_cloudwatch_event_bus_test.go:
```diff
@@ -0,0 +1,184 @@
+func init() {
+	resource.AddTestSweepers("aws_cloudwatch_event_bus", &resource.Sweeper{
+		Name: "aws_cloudwatch_event_bus",
+		F:    testSweepCloudWatchEventBuses,
```
It looks like an event bus can't be deleted if it has associated rules or targets so we'll need to add

```go
Dependencies: []string{""},
```

here.

ewbankkit commented on aws/resource_aws_cloudwatch_event_rule.go:
```diff
@@ -49,6 +50,12 @@ func resourceAwsCloudWatchEventRule() *schema.Resource {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateCloudWatchEventBusName,
```
Maybe don't validate `event_bus_name` here as that would exclude specifying `default` which is a valid event bus name when manipulating events. We'll then need to suppress the difference between `""` and `"default"`.

ewbankkit commented on aws/resource_aws_cloudwatch_event_rule.go:
```diff
@@ -123,7 +130,12 @@ func resourceAwsCloudWatchEventRuleCreate(d *schema.ResourceData, meta interface
+	id := aws.StringValue(input.Name)
+	eventBusName := aws.StringValue(input.EventBusName)
+	if len(eventBusName) > 0 {
+		id = eventBusName + "-" + id
```
I suggest using `/` as the separator as it's not a valid event bus name character while `-` is.
Also don't prefix the name if it's `default` (see https://github.com/terraform-providers/terraform-provider-aws/pull/10256/files#r382815323).

ewbankkit commented on aws/resource_aws_cloudwatch_event_target.go:
```diff
@@ -28,6 +28,13 @@ func resourceAwsCloudWatchEventTarget() *schema.Resource {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateCloudWatchEventBusName,
```
Same as for `aws_cloudwatch_event_rule`.

ewbankkit commented on aws/resource_aws_cloudwatch_event_target.go:
```diff
@@ -255,6 +262,10 @@ func resourceAwsCloudWatchEventTargetCreate(d *schema.ResourceData, meta interfa
 
 	id := rule + "-" + targetId
+	eventBusName := aws.StringValue(input.EventBusName)
+	if len(eventBusName) > 0 {
```
Same as for `aws_cloudwatch_event_rule`.

ewbankkit commented on aws/validators.go:
```diff
@@ -2407,3 +2407,28 @@ func validateRoute53ResolverName(v interface{}, k string) (ws []string, errors [
 	return
 }
+
+func validateCloudWatchEventBusName(v interface{}, k string) (ws []string, errors []error) {
```
Could be simplified (using approach suggested in #8424) to

```go
var validateCloudWatchEventBusName= validation.All(
	validation.StringLenBetween(1, 256),
	validation.StringMatch(regexp.MustCompile(`^[a-zA-Z0-9._\-]+$`), ""),
	StringNotInSlice([]string{"default"}, false), // To implement :)
)
```

ewbankkit commented:
@alexanderkalach Great work. I left some additional comments. The base `aws_cloudwatch_event_bus` is very close. It may be worth splitting that into its own PR with the additional resource changes in their own PRs.
That should make final review and merge quicker.
Thanks.

alexanderkalach commented:
@ewbankkit Thank you for the review. I updated the PR. Also removed the code related to changes of `aws_cloudwatch_event_rule` and `aws_cloudwatch_event_target`. I will create separate PRs for that later.

ewbankkit commented on aws/resource_aws_cloudwatch_event_bus_test.go:
```diff
@@ -0,0 +1,186 @@
+
+func TestAccAWSCloudWatchEventBus_basic(t *testing.T) {
+	var eventBusOutput cloudwatchevents.DescribeEventBusOutput
+	busName := acctest.RandString(15)
```
Minor thing: Prefer `acctest.RandomWithPrefix("tf-acc-test")` so that we can easily distinguish provider acceptance resources in the AWS console.

ewbankkit commented on aws/resource_aws_cloudwatch_event_bus.go:
```diff
@@ -0,0 +1,96 @@
+	_, err := conn.DeleteEventBus(&cloudwatchevents.DeleteEventBusInput{
+		Name: aws.String(d.Id()),
+	})
+	if err != nil {
```
Please add the "resource not found" checking you now have in `resourceAwsCloudWatchEventBusRead()` here also.
Thanks.

ewbankkit commented:
@alexanderkalach Thank for addressing my previous comments.
I left a couple more minor ones.
Verified the acceptance tests:

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSCloudWatchEventBus_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCloudWatchEventBus_ -timeout 120m
=== RUN   TestAccAWSCloudWatchEventBus_basic
=== PAUSE TestAccAWSCloudWatchEventBus_basic
=== RUN   TestAccAWSCloudWatchEventBus_disappears
=== PAUSE TestAccAWSCloudWatchEventBus_disappears
=== CONT  TestAccAWSCloudWatchEventBus_basic
=== CONT  TestAccAWSCloudWatchEventBus_disappears
--- PASS: TestAccAWSCloudWatchEventBus_disappears (19.68s)
--- PASS: TestAccAWSCloudWatchEventBus_basic (40.30s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	41.534s
% TEST=./aws SWEEP=us-west-2 SWEEPARGS=-sweep-run=aws_cloudwatch_event_bus make sweep
WARNING: This will destroy infrastructure. Use only in development accounts.
go test ./aws -v -sweep=us-west-2 -sweep-run=aws_cloudwatch_event_bus -timeout 60m
2020/02/24 09:26:37 [DEBUG] Running Sweepers for region (us-west-2):
2020/02/24 09:26:37 [DEBUG] Sweeper (aws_cloudwatch_event_rule) has dependency (aws_cloudwatch_event_target), running..
2020/02/24 09:26:37 [DEBUG] Running Sweeper (aws_cloudwatch_event_target) in region (us-west-2)
2020/02/24 09:26:37 [INFO] Building AWS auth structure
2020/02/24 09:26:37 [INFO] Setting AWS metadata API timeout to 100ms
2020/02/24 09:26:39 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
2020/02/24 09:26:39 [INFO] AWS Auth provider used: "EnvProvider"
2020/02/24 09:26:39 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/02/24 09:26:39 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/02/24 09:26:40 [DEBUG] Running Sweeper (aws_cloudwatch_event_rule) in region (us-west-2)
2020/02/24 09:26:40 [INFO] Building AWS auth structure
2020/02/24 09:26:40 [INFO] Setting AWS metadata API timeout to 100ms
2020/02/24 09:26:42 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
2020/02/24 09:26:42 [INFO] AWS Auth provider used: "EnvProvider"
2020/02/24 09:26:42 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/02/24 09:26:42 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/02/24 09:26:43 [DEBUG] No CloudWatch Event Rules to sweep
2020/02/24 09:26:43 [DEBUG] Sweeper (aws_cloudwatch_event_target) already ran in region (us-west-2)
2020/02/24 09:26:43 [DEBUG] Sweeper (aws_cloudwatch_event_bus) has dependency (aws_cloudwatch_event_rule), running..
2020/02/24 09:26:43 [DEBUG] Sweeper (aws_cloudwatch_event_rule) has dependency (aws_cloudwatch_event_target), running..
2020/02/24 09:26:43 [DEBUG] Sweeper (aws_cloudwatch_event_target) already ran in region (us-west-2)
2020/02/24 09:26:43 [DEBUG] Sweeper (aws_cloudwatch_event_rule) already ran in region (us-west-2)
2020/02/24 09:26:43 [DEBUG] Sweeper (aws_cloudwatch_event_bus) has dependency (aws_cloudwatch_event_target), running..
2020/02/24 09:26:43 [DEBUG] Sweeper (aws_cloudwatch_event_target) already ran in region (us-west-2)
2020/02/24 09:26:43 [DEBUG] Running Sweeper (aws_cloudwatch_event_bus) in region (us-west-2)
2020/02/24 09:26:43 [INFO] Building AWS auth structure
2020/02/24 09:26:43 [INFO] Setting AWS metadata API timeout to 100ms
2020/02/24 09:26:45 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
2020/02/24 09:26:45 [INFO] AWS Auth provider used: "EnvProvider"
2020/02/24 09:26:45 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/02/24 09:26:45 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/02/24 09:26:46 Sweeper Tests ran successfully:
	- aws_cloudwatch_event_target
	- aws_cloudwatch_event_rule
	- aws_cloudwatch_event_bus
ok  	github.com/terraform-providers/terraform-provider-aws/aws	10.303s
```

alexanderkalach commented:
@ewbankkit I updated the PR according to the last comments. Thanks.

ewbankkit commented:
LGTM.
Acceptance tests:

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSCloudWatchEventBus_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCloudWatchEventBus_ -timeout 120m
=== RUN   TestAccAWSCloudWatchEventBus_basic
=== PAUSE TestAccAWSCloudWatchEventBus_basic
=== RUN   TestAccAWSCloudWatchEventBus_disappears
=== PAUSE TestAccAWSCloudWatchEventBus_disappears
=== CONT  TestAccAWSCloudWatchEventBus_basic
=== CONT  TestAccAWSCloudWatchEventBus_disappears
--- PASS: TestAccAWSCloudWatchEventBus_disappears (17.49s)
--- PASS: TestAccAWSCloudWatchEventBus_basic (36.14s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	37.530s
```

ajfisher commented:
Any update on the current state of this - being able to define non-default event busses in AWS via Terraform would be extremely helpful.

moraneva commented:
Wanted to check in a month after the last comment, has there been any update on getting this merged in? We are currently having to use a combination of cloud formation and terraform in order to manage EventBridge resources

alexanderkalach commented:
I've created PRs for [aws_cloudwatch_event_rule](https://github.com/terraform-providers/terraform-provider-aws/pull/12886) and [aws_cloudwatch_event_target](https://github.com/terraform-providers/terraform-provider-aws/pull/12887). It should be enough to close https://github.com/terraform-providers/terraform-provider-aws/issues/9330 
@ewbankkit could you review other PRs when you have time?
Thanks

ewbankkit commented:
@alexanderkalach Thanks for the additional PRs.
Please see [this comment](https://github.com/terraform-providers/terraform-provider-aws/issues/9330#issuecomment-616884604); There is progress on getting the Event Bridge functionality merged.

teamterraform commented:
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

gdavison reviewed (approved):
LGTM 🚀 

--- PASS: TestAccAWSCloudWatchEventBus_default (17.28s)
--- PASS: TestAccAWSCloudWatchEventBus_disappears (46.43s)
--- PASS: TestAccAWSCloudWatchEventBus_basic (113.36s)
--- PASS: TestAccAWSCloudWatchEventBus_tags (139.77s)

barneyparker commented:
Fantastic!!!   Great work @alexanderkalach @ewbankkit @gdavison - Your work on this is really very, very appreciated :+1:
